diff -u mpg123-0.59o/buffer.c mpg123-0.59op/buffer.c
--- mpg123-0.59o/buffer.c	Wed Feb 11 13:55:18 1998
+++ mpg123-0.59op/buffer.c	Mon Feb 16 11:57:29 1998
@@ -15,6 +15,7 @@
 
 static int intflag = FALSE;
 static int usr1flag = FALSE;
+static int stopped = FALSE;
 
 static void catch_interrupt (void)
 {
@@ -26,6 +27,19 @@
 	usr1flag = TRUE;
 }
 
+static void catch_usr2 (void)
+{
+  stopped = TRUE;
+  
+  while(stopped)
+    sleep(60000);
+}
+
+static void catch_cont(void)
+{
+  stopped = FALSE;
+}
+
 #if !defined(OS2) && !defined(GENERIC) && !defined(WIN32)
 
 void buffer_loop(struct audio_info_struct *ai, sigset_t *oldsigset)
@@ -37,6 +51,9 @@
 
 	catchsignal (SIGINT, catch_interrupt);
 	catchsignal (SIGUSR1, catch_usr1);
+	catchsignal (SIGUSR2, catch_usr2);
+	catchsignal (SIGCONT, catch_cont);
+	
 	sigprocmask (SIG_SETMASK, oldsigset, NULL);
 	if (param.outmode == DECODE_AUDIO) {
 		if (audio_open(ai) < 0) {
Common subdirectories: mpg123-0.59o/jukebox and mpg123-0.59op/jukebox
diff -u mpg123-0.59o/mpg123.c mpg123-0.59op/mpg123.c
--- mpg123-0.59o/mpg123.c	Wed Feb 11 14:15:08 1998
+++ mpg123-0.59op/mpg123.c	Mon Feb 16 12:20:57 1998
@@ -60,6 +60,7 @@
 
 static int intflag = FALSE;
 static int remflag = FALSE;
+static int stopped = FALSE;
 
 #ifndef _WIN32
 static void catch_child(void)
@@ -71,6 +72,25 @@
 {
   intflag = TRUE;
 }
+
+static void catch_cont(void)
+{
+  stopped = FALSE;
+}
+
+static void catch_usr2(void)
+{
+  stopped = TRUE;
+  fprintf(stderr, "mpg123 Stopped\n");
+  if(param.usebuffer)
+    kill(buffer_pid, SIGUSR2);
+    
+  while (stopped)
+    sleep(60000);
+    
+  if (param.usebuffer)
+    kill(buffer_pid, SIGCONT);   
+}
 #endif
 
 static char remote_buffer[1024];
@@ -749,6 +769,8 @@
 
 #ifndef WIN32
 	catchsignal (SIGINT, catch_interrupt);
+	catchsignal (SIGUSR2, catch_usr2);
+	catchsignal (SIGCONT, catch_cont);
 
 	if(frontend_type) {
 		handle_remote();
@@ -801,12 +823,12 @@
 	if(param.usebuffer) {
 		int s;
         	struct timeval wait100 = {0, 100000};
-		while ((s = xfermem_get_usedspace(buffermem))) {
+/*		while ((s = xfermem_get_usedspace(buffermem))) {
 			if(buffermem->wakeme[XF_READER] == TRUE)
 				break;
 			print_stat(&fr,frameNum,xfermem_get_usedspace(buffermem),&ai);
-                	select(1, NULL, NULL, NULL, &wait100);
-		}
+                	select(1, NULL, NULL, NULL, &wait100); 
+		} */
 	}
 
 	if (!param.quiet) {
@@ -838,12 +860,18 @@
         if (secdiff < 1000)
           break;
 #endif
-        intflag = FALSE;
       }
     }
 #if !defined(OS2) && !defined(GENERIC) && !defined(WIN32)
     if (param.usebuffer) {
-      xfermem_done_writer (buffermem);
+      if (intflag)
+        {
+          fprintf(stderr, "Killing buffer PID...\n");
+          kill(buffer_pid, SIGKILL);
+          intflag = FALSE;
+        }
+      else  
+        xfermem_done_writer (buffermem);        
       waitpid (buffer_pid, NULL, 0);
       xfermem_done (buffermem);
     }
Common subdirectories: mpg123-0.59o/plugin and mpg123-0.59op/plugin
